src/debug/ee/frameinfo.cpp

```diff
@@ -1559,20 +1559,27 @@ StackWalkAction DebuggerWalkStackProc(CrawlFrame *pCF, void *data)
     }
     else
 #endif // FEATURE_EH_FUNCLETS
-    // We should ignore IL stubs with no frames in our stackwalking.
-    // The only exception is dynamic methods.  We want to report them when SIS is turned on.
+    // We ignore most IL stubs with no frames in our stackwalking. As exceptions
```

Comment:
What testing has been done so far? We've now exposed the tailcall stub in all stackwalks the debugger does for any purpose and the logic which interacts with those stackwalk results presumably wasn't written with the tailcall stub in mind originally.

Reply:
I have done only rudimentary debugging -- confirmed that stepping in, stepping over and stepping out works properly. I did search for where the other special case (multicast stubs) is handled, however, and it is only handled in that one place where I added my own special case.

The special case for the `CallTailCallTarget` stub is actually only there to allow the debugger to be able to implement the `IsTailCallThatReturns` function in controller.cpp. Without exposing the stack frame, `GetControlPC(&info->m_returnFrame.registers)` will not actually be the return PC -- instead, it will be to some other frame further up the stack. However I specifically need to be able to check the direct return address.

A better solution might be to do a "clean" stack walk in that function by using the stack walking function on the debugger interface. I tried this but I was not able to get it to work after a while. There are a bunch of terms about each frame that I am unfamiliar with and that the debugger handles specially during stack walks. For instance, even the previous comment does not make sense to me -- how do these IL stubs not have frames?

Perhaps there is an easier way to get the "direct" return address. In any case I am open for input.

Reply:
> I have done only rudimentary manual testing -- confirmed that stepping in, stepping over and stepping out works properly

Given the relative rarity of tailcall and our early point in the .NET 5 development cycle I think this is a sufficient level of testing for it. Thanks!

> For instance, even the previous comment does not make sense to me -- how do these IL stubs not have frames?

When you see code or comments that suggests that some stack frames don't have frames (it sounds non-sensical I know) there are two different notions of 'frame' being used. The runtime has a class called "Frame" and some methods will store an instance of the "Frame" class on the stack. Every method call has the normal stack frame where registers are saved, local variables are stored, etc. So the comment really means that a given method call did not initialize a "Frame" object on the stack. Ideally we refer to clr's Frame type always using a capitol 'F', but sometimes you have to guess based on context which meaning is being used.

Reply:
Thanks for the explanation, that makes sense.